retrofit-adapters/rxjava2/src/main/java/retrofit2/adapter/rxjava2/RxJava2CallAdapter.java

```diff
@@ -83,6 +84,6 @@
     if (isCompletable) {
       return observable.ignoreElements();
     }
-    return observable;
+    return RxJavaPlugins.onAssembly(observable);
```

Comment:
Hm, I thought only custom observable instantiation will be wrapped:

```java
Observable<Response<R>> responseObservable = RxJavaPlugins.onAssembly(isAsync
        ? new CallEnqueueObservable<>(call)
        : new CallExecuteObservable<>(call));
```

```java
if (isResult) {
  observable = RxJavaPlugins.onAssembly(new ResultObservable<>(responseObservable));
} else if (isBody) {
  observable = RxJavaPlugins.onAssembly(new BodyObservable<>(responseObservable));
} else {
  observable = responseObservable;
}
```

Current implementation is not bad, but:

- As far as I see stactrace won't point to specific lines where `CallEnqueueObservable`/`CallExecuteObservable`/`ResultObservable`/`BodyObservable` got intantiated but will only point to line 87 where it can be hard to figure which branch of code was executed and what is the original observable type if `subscribeOn` was applied
- Wrapping happens after potential `subscribeOn`. Retrofit Observable types are package-private which means that reflecting on their internals is not part of public api provided by Retrofit, however it seems legal to check for `package` of the Observable type to apply some custom behavior. Since wrapping happens after `subscribeOn`, type information gets hidden.

Reply:
I've made a branch with more fine-grained wrapping that we can turn into PR if you want 

https://github.com/square/retrofit/compare/master...artem-zinnatullin:az/rxjava-adapter-onAssembly-fix?expand=1